### PR TITLE
feat: add basic SQL features

### PR DESCRIPTION
**Pull request by eitsupi**
There is no `register_globals` setting present in py-polars for now.

By the way, I would like to create a `DBI` backend based on this feature out of interest, but is it okay to create a repository in the rpolars organization? (It is more convenient to exist in the same R-universe)

cc @paleolimbot

### CODE REVIEW

**Comment by eitsupi** [comment]
Oops, it seems that this feature can only be built using the nightly-toolchain...

**Comment by etiennebacher** [comment]
> Oops, it seems that this feature can only be built using the nightly-toolchain...

Should I review now?

**Comment by eitsupi** [comment]
> Should I review now?

Thanks for checking this.
Yes, the problem should be fixed.......

File: NEWS.md [hunk 1]
```diff
@@ -21,6 +21,11 @@
   - `$scan_parquet()` gains an argument `hive_partitioning`.
   - `$meta$tree_format()` has a better formatted output.
 
+## What's changed
+
+- New class `RPolarsSQLContext` and its methods.
```
Comment (etiennebacher) [inline NEWS.md:26]: Can you add some information on what it allows to do? Something like: 
```suggestion
- New class `RPolarsSQLContext` and its methods to perform any SQL query on a Polars DataFrame or LazyFrame.
```
(not sure what I added is correct)

File: R/sql.R [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+#' @examplesIf pl$polars_info()$features$sql
+#' lf = pl$LazyFrame(a = 1:3, b = c("x", NA, "z"))
+#' res = pl$SQLContext(frame = lf)$execute(
+#'   "SELECT b, a*2 AS two_a FROM frame WHERE b IS NOT NULL"
+#' )
+#' res$collect()
+RPolarsSQLContext
+
+
+#' @title auto complete $-access into a polars object
```
Comment (etiennebacher) [inline R/sql.R:16]:
```suggestion
#' @title Auto complete $-access into a polars object
```

File: R/sql.R [hunk 3]
```diff
@@ -0,0 +1,167 @@
+#' @title Run SQL queries against DataFrame/LazyFrame data.
```
Comment (etiennebacher) [inline R/sql.R:1]:
```suggestion
#' Run SQL queries against DataFrame/LazyFrame data.
```
This tag is not actually needed

File: R/sql.R [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+#' lf = pl$LazyFrame(a = 1:3, b = c("x", NA, "z"))
+#' res = pl$SQLContext(frame = lf)$execute(
+#'   "SELECT b, a*2 AS two_a FROM frame WHERE b IS NOT NULL"
+#' )
+#' res$collect()
+RPolarsSQLContext
+
+
+#' @title auto complete $-access into a polars object
+#' @description called by the interactive R session internally
```
Comment (etiennebacher) [inline R/sql.R:17]:
```suggestion
#' @description Called by the interactive R session internally
```

File: R/sql.R [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+#' @param x RPolarsSQLContext
+#' @param pattern code-stump as string to auto-complete
```
Comment (etiennebacher) [inline R/sql.R:19]: what's a code-stump?

File: R/sql.R [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+#' @return char vec
```
Comment (etiennebacher) [inline R/sql.R:20]:
```suggestion
#' @return Character vector
```
Can you clean the rest?

File: docs/make-docs.R [hunk 7]
```diff
@@ -136,7 +136,7 @@ make_doc_hierarchy = function() {
     "pl", "Series", "DataFrame", "LazyFrame", "GroupBy",
     "LazyGroupBy", "ExprList", "ExprBin", "ExprCat", "ExprDT",
     "ExprMeta", "ExprName", "ExprStr", "ExprStruct",
-    "Expr", "IO", "RThreadHandle"
+    "Expr", "IO", "RThreadHandle", "SQLContext"
```
Comment (etiennebacher) [inline docs/make-docs.R:139]: "SQL" would be clearer

**Review by etiennebacher — APPROVED** [review]
Looks cool, thanks! I personally don't use SQL but there are probably many people who will be happy to see this feature.

Mostly comments about docs

**Review by etiennebacher — CHANGES_REQUESTED** [review]
actually tests fail so I'm "unapproving" for now

File: docs/make-docs.R (hunk 7 above)
Comment (eitsupi) [reply docs/make-docs.R:139]: I am not sure how this works, I assumed that SQLContext is the name of the actual structure used upstream and here it is related to the name of the Rd file.

File: R/sql.R (hunk 5 above)
Comment (eitsupi) [reply R/sql.R:19]: I believe this is copied from elsewhere.

File: R/sql.R (hunk 6 above)
Comment (eitsupi) [reply R/sql.R:20]: ditto

File: R/sql.R (hunk 4 above)
Comment (eitsupi) [reply R/sql.R:17]: I believe this is copied from elsewhere.

File: R/sql.R (hunk 3 above)
Comment (eitsupi) [reply R/sql.R:1]: I know it is, but I used it because it appeared to be used elsewhere.

**Comment by eitsupi** [comment]
Thanks for your review.
Since `.DollarNames.RPolarsSQLContext` is completely copied from elsewhere, it would be better to revise it with other places instead of this PR.

File: docs/make-docs.R (hunk 7 above)
Comment (etiennebacher) [reply docs/make-docs.R:139]: Yeah, I think the change should happen below actually, where we rename the Expr variants

Comment (etiennebacher) [reply docs/make-docs.R:139]: Not even that, forget about this comment, I'll take care of this later

**Comment by etiennebacher** [comment]
> By the way, I would like to create a DBI backend based on this feature out of interest, but is it okay to create a repository in the rpolars organization?

I don't know but you should open an issue about that

**Review by etiennebacher — APPROVED** [review]

**Pull request merged**